### DISCUSSION TITLE
Module import fails when creating and importing from Rust

### DISCUSSION THREAD

**Discussion by flickpp** [PyO3/pyo3#2649] [Questions] [answered]
This is probably just a documentation issue, I can't find the answer on pyo3.rs.

This code:

```rust
use pyo3::prelude::*;

fn main() {
    Python::with_gil(|py| -> PyResult<()> {
        let mod1 = PyModule::new(py, "my_module")?;
        let mod2 = PyModule::import(py, "my_module")?;

        Ok(())
    })
    .expect("didn't import module");
}
```

produces
```
thread 'main' panicked at 'didn't import module: PyErr { type: <class 'ModuleNotFoundError'>, value: ModuleNotFoundError("No module named 'my_module'"), traceback: None }', src/main.rs:10:6
```

**what I expected to happen**
mod2 should be a reference to the python module.

**what I'm trying to do**
I'm trying to call Python code from inside Rust, which itself calls back into Rust.
Roughly I'm trying to do this.

```rust
use pyo3::prelude::*;

fn main() {
    Python::with_gil(|py| -> PyResult<()> {
        let mod1 = PyModule::new(py, "my_module")?;
        py.run("import my_module", None, None)?;

        Ok(())
    })
    .expect("didn't import module");
}
```

**Suggestion**
On the "calling Python from rust" section of pyo3.rs, add either an example or some docs on how/why this won't work.

**Comment by birkenfeld** [comment, marked as answer]
Creating a new module does not make it known to the import machinery. You'll also have to add it to the module registry (`sys.modules`) since `import` will not find it in the usual way (looking for files on `sys.path`).

**Reply by flickpp** [reply]
Thanks! Works as expected.

I get this is obvious to people who know Python internals, but is it worth adding something to the docs?

**Reply by mejrs** [reply]
I agree.

It's in the maturin starter example (https://github.com/PyO3/pyo3/blob/main/examples/maturin-starter/src/lib.rs#L30), but I wouldn't expect someone to sift through the examples. It'd be nice to mention it somewhere on the pymodule page

**Reply by flickpp** [reply]
If this done come to pass - I'd like to volunteer myself.

**Reply by mejrs** [reply]
If you're willing to write it up that would be appreciated :)

**Reply by birkenfeld** [reply]
Looks like there is a valid issue here, sorry for converting it prematurely.
